File: q51-75/q54.py
# 2020.08.22
# Problem Statement:
# https://leetcode.com/problems/spiral-matrix/

class Solution:
    # helper function for appending each round
    def spiralOrderHelper(self, matrix, start_row, start_col, answer):
        
        # if only contains a row, append the row
        if start_row+1 == len(matrix) - start_row: 
            for i in range(start_col, len(matrix[0]) - start_col):
                answer.append(matrix[start_row][i])
            return answer
        
        # if only contains a col, append the col
        elif start_col+1 == len(matrix[0]) - start_col:
            for i in range(start_row, len(matrix) - start_row):
                answer.append(matrix[i][start_col])
            return answer
        
        # otherwise
        # append the top row
        for i in range(start_col, len(matrix[0]) - start_col):
            answer.append(matrix[start_row][i])
        
        # append the right col
        for i in range(start_row+1, len(matrix) - start_row):
            answer.append(matrix[i][len(matrix[0]) - start_col - 1])
        
        # append the bottom row
        for i in range(start_col+1, len(matrix[0]) - start_col):
            answer.append(matrix[len(matrix) - start_row - 1][len(matrix[0])-1-i])

        # append the left col    
        for i in range(start_row+1, len(matrix) - start_row - 1):
            answer.append(matrix[len(matrix)-1-i][start_col])
        
        return answer
        
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        # check empty case
        if matrix == []: return []
        
        # initialize answer to return, and the top-left position of a round
        answer = []
        row, col = 0, 0
        
        # should not go beyond min(len(matrix[0])-1, len(matrix)-1) to avoid duplication
        x = min(len(matrix[0])-1, len(matrix)-1)
        while row <= x // 2:
            answer = self.spiralOrderHelper(matrix, row, col, answer)
            row = row + 1
            col = col + 1
        return answer
        

File: Amazon/q54.java
// 2022.06.03
// Problem Statement:
// https://leetcode.com/problems/spiral-matrix/

// idea: from external circle to internal circle, if the smaller side is odd, 
// need to append an extra line (vertical or horizontal)
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> answer = new ArrayList<> ();
        int m=matrix.length, n=matrix[0].length;
        for (int i=0; i<(Math.min(m, n)+1)/2; i++) { // circle
            if (answer.size()==m*n) {
                break;
            }
            if (m==n && 2*i+1==n) {
                answer.add(matrix[m/2][n/2]);
                break;
            }
            if (m>n && i==(n+1)/2-1 && n%2!=0) { // vertical line
                for (int x=i; x<m-i; x++) {
                    answer.add(matrix[x][i]);
                }
                break;
            }
            if (m<n && i==(m+1)/2-1 && m%2!=0) { // horizontal line
                for (int y=i; y<n-i; y++) {
                    answer.add(matrix[i][y]);
                }
                break;
            }
            // top
            for (int col=i; col<n-i-1; col++) {
                answer.add(matrix[i][col]);
            }
            // right
            for (int row=i; row<m-i-1; row++) {
                answer.add(matrix[row][n-1-i]);
            }
            // bottom
            for (int col=n-i-1; col>i; col--) {
                answer.add(matrix[m-1-i][col]);
            }
            // left
            for (int row=m-i-1; row>i; row--) {
                answer.add(matrix[row][i]);
            }
        }
        return answer;
    }
}